File: go/stack_03/array_stack.go
package stack_03

import (
	"fmt"
)

type ArrayStack struct {
	top  int
	data [] interface{} //	 存放数据
}

func NewArrayStack() *ArrayStack {
	return &ArrayStack{
		top:  -1,
		data: make([] interface{}, 0, 8),
	}
}

func (s *ArrayStack) IsEmpty() bool {
	return s.top == -1
}

func (s *ArrayStack) Push(value interface{}) {

	if len(s.data)+1 <= cap(s.data) {
		fmt.Println("入栈")
		s.top += 1
		s.data = append(s.data, value)

	} else {
		fmt.Println("栈已满，无法入栈")
	}

}

func (s *ArrayStack) Pop() interface{} {
	if s.IsEmpty() {
		fmt.Print("栈已空，无法出栈")
		return nil
	} else {
		fmt.Println("出栈")
		v := s.data[s.top]
		s.top -= 1
		return v

	}
}

func (s *ArrayStack) Print() {
	if s.IsEmpty() {
		fmt.Println("空栈")
	} else {
		fmt.Println("栈中剩余数据，出栈顺序为：")
		for i :=s.top ; i >= 0; i-- {
			fmt.Println(s.data[i])
		}
	}
}

File: python/03_stack/array_stack.py
# -*- coding: utf-8 -*-

class ArrayStack:
    def __init__(self):
        self.stack = []
        self.size = 0

    def push(self, value):
        self.stack.append(value)
        self.size += 1

    def pop(self):
        if self.size > 0:
            self.size -= 1
            return self.stack.pop()

    @property
    def top(self):
        if self.stack:
            return self.stack[-1]

    def is_empty(self):
        return self.size == 0

    def __repr__(self):
        return str(self.stack)


if __name__ == "__main__":
    stack = ArrayStack()
    for i in range(9):
        stack.push(i)
    print(stack)
    for _ in range(3):
        stack.pop()
    print(stack)
